src/terminal.ts

```diff
@@ -106,6 +131,45 @@ export class BrowserTerminal implements vscode.Pseudoterminal {
         }
     }
 
+    removeQuestionMark() {
+        // use this to indicate autocomplete attempt
+        // TODO: make this some emoji type character that users would never use
+        if (this.currentLine.endsWith("?")) {
+            this.removeNCharacters(1);
+        }
+    }
+    
+    addSnippetToCurrentLine(snippet: string) {
+        this.currentLine += snippet;
+        this.writeEmitter.fire(snippet);
+    }
+
+    clearCurrentCommand() {
+        // remove last character
+        const command = this.currentLine.slice(this.constructPrompt().length);
+
+        this.removeNCharacters(command.length);
+    }
+
+    removeNCharacters(n: number) {
+        for (let i = 0; i < n; i++) {
+            this.currentLine = this.currentLine.substring(0, this.currentLine.length - 1);
+            this.writeEmitter.fire(ACTIONS.cursorBack);
+            this.writeEmitter.fire(ACTIONS.deleteChar);
+        }
+    }
+
+    getHistory(historyIndex: number): string {
+        // if no more history
+        if (historyIndex < 0 
+            || this.commandHistory.length == 0
+            || historyIndex >= this.commandHistory.length) {
+            this.historyPointer = this.commandHistory.length;
+            return "----";
```

Comment:
This will print out ---- in the terminal right? It might be annoying to have to backspace that away.. 

Looking at bash it seems to:
- for historyIndex < 0 it just keeps the item at historyIndex == 0 and doesnt change it and that indicates end of history to the user
- for historyIndex >= commandHistory.length  it returns the line back to what it was before - ie effectively even before command is executed it's added to history and locks it at the end.

Reply:
Good point will fix!